serverutils: prevent start tenant when probabilistic behaviour is enabled

herkolategan opened the pull request:
The probabilistic behaviour that maybe starts a test tenant currently present in
test server and test cluster is not widely known or apparent. This change seeks
to alleviate that by warning, via an error, new test authors that if they start
a tenant in a test there is the possibility that a default test tenant could
also be started. This is so that the author of a test can make an informed
decision whether to control the starting of tenants or allow probabilistic
behaviour to take effect.

Previously there was a `DisableDefaultTestTenant` test arg for tests that wanted
to opt out of the probabilistic behaviour. This argument has been augmented to
now support multiple modes for more flexibility. Tests can now choose to rely on
the probabilistic test tenant, or create its own tenants, or both. The main
purpose of the default mode which disallows the starting of tenants is to
enlighten test authors of the possible presence of the default test tenant. It
is still possible to have the probabilistic behaviour and start tenants in the
test as well, but it has to be explicitly set in the arguments if the test
author seeks this behaviour.

Numerous tests that currently start the test tenant in the test itself, had this
behaviour. Those tests have been updated to disable the default test tenant from
probabilistically starting. As these tests would now have raised an error
(warning) without changes.

Epic: CRDB-18499

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98562)
<!-- Reviewable:end -->

herkolategan committed:
serverutils: prevent start tenant when probabilistic behaviour is active

The probabilistic behaviour that maybe starts a test tenant currently present in
test server and test cluster is not widely known or apparent. This change seeks
to alleviate that by warning, via an error, new test authors that if they start
a tenant in a test there is the possibility that a default test tenant could
also be started. This is so that the author of a test can make an informed
decision whether to control the starting of tenants or allow probabilistic
behaviour to take effect.

Previously there was a `DisableDefaultTestTenant` test arg for tests that wanted
to opt out of the probabilistic behaviour. This argument has been augmented to
now support multiple modes for more flexibility. Tests can now choose to rely on
the probabilistic test tenant, or create its own tenants, or both. The main
purpose of the default mode which disallows the starting of tenants is to
enlighten test authors of the possible presence of the default test tenant. It
is still possible to have the probabilistic behaviour and start tenants in the
test as well, but it has to be explicitly set in the arguments if the test
author seeks this behaviour.

Numerous tests that currently start the test tenant in the test itself, had this
behaviour. Those tests have been updated to disable the default test tenant from
probabilistically starting. As these tests would now have raised an error
(warning) without changes.

Epic: CRDB-18499

knz reviewed (commented):
This is excellent work. It must have been a lot of effort too.
Just a few things and this will soon be good to merge.

> Reviewed 111 of 111 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-:-NRY21lODtBxBfNocl2m:b-59mo7c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @herkolategan, @miretskiy, @renatolabs, and @smg260)

___
*[`pkg/sql/logictest/logic.go` line 1319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRXyt1K6ZUEVYIvv5-T:-NRXyt1K6ZUEVYIvv5-U:b-1pllno) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/sql/logictest/logic.go#L1319)):*
> ```Go
> 
> 	var defaultTestTenant base.DefaultTestTenantOptions
> 	if t.cfg.UseTenant || t.cfg.DisableDefaultTestTenant {
> ```

Do we want to keep separate fields `UseTenant` and `DisableDefaultTestTenant` side-by-side? Can't we replace them with a single field of type `DefaultTestTenantOptions` too?
___
*[`pkg/testutils/testcluster/testcluster.go` line 457 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRXzR9v9L-nt8hMkUJt:-NRXzR9v9L-nt8hMkUJu:bxduk8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/testutils/testcluster/testcluster.go#L457)):*
> ```Go
> 	})
> 
> 	if preventFurtherTenants {
> ```

This logic should be in `AddServer()` not here. Otherwise you don't get it configured properly for servers created later.
___
*[`pkg/ccl/cloudccl/externalconn/datadriven_test.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRXzm_s6FUsIEGDoXF7:-NRXzm_s6FUsIEGDoXF8:b-9ihdzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/ccl/cloudccl/externalconn/datadriven_test.go#L54)):*
> ```Go
> 		tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
> 			ServerArgs: base.TestServerArgs{
> 				DefaultTestTenant: base.TestTenantProbabilistic,
> ```

Why not the default value here?
___
*[`pkg/ccl/serverccl/server_sql_test.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRY-sYy4PJ0t0ACcyer:-NRY-sYy4PJ0t0ACcyes:b3yu65a) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/ccl/serverccl/server_sql_test.go#L55)):*
> ```Go
> 
> 	tc := serverutils.StartNewTestCluster(t, 3, base.TestClusterArgs{ServerArgs: base.TestServerArgs{
> 		// We need to disable the default test tenant because we're going to create our.
> ```

nit: "our own."
___
*[`pkg/ccl/testccl/sqlccl/temp_table_clean_test.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRY0iEaCjKjncTsMFIv:-NRY0iEbCphuoKTbkavi:bcsupu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/ccl/testccl/sqlccl/temp_table_clean_test.go#L98)):*
> ```Go
> 		t, 3 /* numNodes */, base.TestClusterArgs{ReplicationMode: base.ReplicationManual,
> 			ServerArgs: base.TestServerArgs{
> 				// Disable the default test tenant so that we can start it
> ```

nit: period at end of sentence.


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98562)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @knz, @miretskiy, @renatolabs, and @smg260)

___
*[`pkg/sql/logictest/logic.go` line 1319 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRXyt1K6ZUEVYIvv5-T:-NRZeHlYAf9RbelWThfQ:b-vg43e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/sql/logictest/logic.go#L1319)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Do we want to keep separate fields `UseTenant` and `DisableDefaultTestTenant` side-by-side? Can't we replace them with a single field of type `DefaultTestTenantOptions` too?
</blockquote></details>

Well spotted, these two can be combined. Although there is a slight twist on the logic that if a tenant is explicitly requested we allow `logic test` to handle the creation of the tenant rather than test server / cluster. I left it that way since `logic test` has some of its own logic tied to explicitly tenant creation.
___
*[`pkg/testutils/testcluster/testcluster.go` line 457 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRXzR9v9L-nt8hMkUJt:-NRbEamIChrgk5qZ7qmc:bxz5jeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/testutils/testcluster/testcluster.go#L457)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This logic should be in `AddServer()` not here. Otherwise you don't get it configured properly for servers created later.
</blockquote></details>

True, this does not cover all cases, I'll update.
___
*[`pkg/ccl/cloudccl/externalconn/datadriven_test.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRXzm_s6FUsIEGDoXF7:-NRbLkDd3GBwb8BZIn1x:b831wxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/ccl/cloudccl/externalconn/datadriven_test.go#L54)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Why not the default value here?
</blockquote></details>

There is a case here where a tenant is started manually by the test: `externalConnTestCluster.InitializeTenant(ctx, tenantID)`. So here the default is not chosen as this test can start its own tenants.



_Code quote:_
```Go
DefaultTestTenant: base.TestTenantProbabilistic,
```
___
*[`pkg/ccl/serverccl/server_sql_test.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRY-sYy4PJ0t0ACcyer:-NRbEgez7DOCdJzHhnSf:b-39v3cp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/ccl/serverccl/server_sql_test.go#L55)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: "our own."
</blockquote></details>

Whoops, fixed.
___
*[`pkg/ccl/testccl/sqlccl/temp_table_clean_test.go` line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-NRY0iEaCjKjncTsMFIv:-NRbEiqVARxiq5LUeN9g:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf54e863bb221e9956fa576e2bc4fee6d696a3e5/pkg/ccl/testccl/sqlccl/temp_table_clean_test.go#L98)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: period at end of sentence.
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

knz commented:
@herkolategan CI is telling you something

herkolategan commented:
> @herkolategan CI is telling you something

Thanks, snuck in with a rebase. I'll update.

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 19 of 32 files at r2, 5 of 7 files at r3, 19 of 19 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98562#-:-NS5IntSBdnf1Gxo6qoS:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @benbardin, @miretskiy, @renatolabs, and @smg260)



<!-- Sent from Reviewable.io -->

herkolategan commented:
bors r=knz
